File: backend/tests/unit/test_services.py
import pytest
from unittest.mock import AsyncMock, MagicMock, patch
from backend.services.agent_service import AgentService
from backend.services.llm_service import LLMService
from backend.models.agent import Agent, AgentCreate, AgentStatus, AgentCapability

@pytest.fixture
def mock_db():
    """Create a mock database"""
    db = MagicMock()
    db.fetch_one = AsyncMock()
    db.fetch_all = AsyncMock()
    db.execute = AsyncMock()
    return db

@pytest.mark.asyncio
async def test_agent_service_get_agent(mock_db):
    """Test getting an agent by ID"""
    # Mock database response
    mock_db.fetch_one.return_value = {
        "id": "test-agent-1",
        "name": "Test Agent",
        "description": "A test agent",
        "model": "test-model",
        "capabilities": '["web_research", "code_generation"]',
        "config": '{"temperature": 0.7}',
        "status": "active",
        "created_at": "2023-01-01T00:00:00",
        "updated_at": "2023-01-01T00:00:00"
    }
    
    service = AgentService()
    service.db = mock_db
    
    agent = await service.get_agent("test-agent-1")
    
    assert agent is not None
    assert agent.id == "test-agent-1"
    assert agent.name == "Test Agent"
    assert AgentCapability.WEB_RESEARCH in agent.capabilities
    assert agent.status == AgentStatus.ACTIVE

@pytest.mark.asyncio
async def test_llm_service_generate_response():
    """Test generating a response from LLM service"""
    service = LLMService()
    
    # Mock the API calls
    with patch('aiohttp.ClientSession.post') as mock_post:
        mock_response = AsyncMock()
        mock_response.status = 200
        mock_response.json.return_value = {
            "choices": [{"message": {"content": "Test response"}}]
        }
        mock_post.return_value.__aenter__.return_value = mock_response
        
        response = await service.generate_response("test-agent", "Hello")
        
        assert "Test response" in response

@pytest.mark.asyncio
async def test_llm_service_get_available_models():
    """Test getting available models"""
    service = LLMService()
    
    # Mock the API calls for both LM Studio and Ollama
    with patch('aiohttp.ClientSession.get') as mock_get:
        # Mock LM Studio response
        mock_response_lm = AsyncMock()
        mock_response_lm.status = 200
        mock_response_lm.json.return_value = {
            "data": [{"id": "test-model-1"}]
        }
        
        # Mock Ollama response  
        mock_response_ollama = AsyncMock()
        mock_response_ollama.status = 200
        mock_response_ollama.json.return_value = {
            "models": [{"name": "test-model-2"}]
        }
        
        # Make mock_get return different responses based on URL
        def side_effect(*args, **kwargs):
            url = args[0]
            if "lmstudio" in url:
                return mock_response_lm
            else:
                return mock_response_ollama
        
        mock_get.side_effect = side_effect
        
        models = await service.get_available_models()
        
        # Should have models from both services
        assert len(models) >= 2
        assert any(m["type"] == "lmstudio" for m in models)
        assert any(m["type"] == "ollama" for m in models)

File: test/backend/unit/test_services.js
# Backend Services Unit Tests
import pytest
from unittest.mock import Mock, patch
from backend.services.agent_service import AgentService
from backend.services.llm_service import LLMService
from backend.services.workflow_service import WorkflowService
from backend.models.agent import Agent
from backend.models.workflow import Workflow

class TestAgentService:
    def test_create_agent(self):
        """Test creating a new agent"""
        service = AgentService()
        agent_data = {
            "name": "Test Agent",
            "description": "A test agent",
            "model": "lm_studio",
            "capabilities": ["research", "analysis"]
        }
        
        agent = service.create_agent(agent_data)
        
        assert agent.id is not None
        assert agent.name == "Test Agent"
        assert agent.description == "A test agent"
        assert agent.model == "lm_studio"
        assert "research" in agent.capabilities
        assert "analysis" in agent.capabilities
    
    def test_get_agent(self):
        """Test retrieving an agent by ID"""
        service = AgentService()
        agent_data = {
            "name": "Test Agent",
            "description": "A test agent",
            "model": "lm_studio"
        }
        
        created_agent = service.create_agent(agent_data)
        retrieved_agent = service.get_agent(created_agent.id)
        
        assert retrieved_agent is not None
        assert retrieved_agent.id == created_agent.id
        assert retrieved_agent.name == created_agent.name
    
    def test_update_agent(self):
        """Test updating an existing agent"""
        service = AgentService()
        agent_data = {
            "name": "Test Agent",
            "description": "A test agent",
            "model": "lm_studio"
        }
        
        agent = service.create_agent(agent_data)
        update_data = {"description": "Updated description"}
        
        updated_agent = service.update_agent(agent.id, update_data)
        
        assert updated_agent.description == "Updated description"
        assert updated_agent.name == "Test Agent"  # Should remain unchanged
    
    def test_delete_agent(self):
        """Test deleting an agent"""
        service = AgentService()
        agent_data = {
            "name": "Test Agent",
            "description": "A test agent",
            "model": "lm_studio"
        }
        
        agent = service.create_agent(agent_data)
        assert service.get_agent(agent.id) is not None
        
        service.delete_agent(agent.id)
        assert service.get_agent(agent.id) is None

class TestLLMService:
    @patch('backend.services.llm_service.requests.post')
    def test_send_message_to_lm_studio(self, mock_post):
        """Test sending a message to LM Studio"""
        mock_response = Mock()
        mock_response.json.return_value = {
            "choices": [{"message": {"content": "Test response from LM Studio"}}]
        }
        mock_response.raise_for_status.return_value = None
        mock_post.return_value = mock_response
        
        service = LLMService()
        response = service.send_message("lm_studio", "Test message", {})
        
        assert response == "Test response from LM Studio"
        mock_post.assert_called_once()
    
    @patch('backend.services.llm_service.requests.post')
    def test_send_message_to_ollama(self, mock_post):
        """Test sending a message to Ollama"""
        mock_response = Mock()
        mock_response.json.return_value = {"response": "Test response from Ollama"}
        mock_response.raise_for_status.return_value = None
        mock_post.return_value = mock_response
        
        service = LLMService()
        response = service.send_message("ollama", "Test message", {})
        
        assert response == "Test response from Ollama"
        mock_post.assert_called_once()
    
    def test_handle_llm_error(self):
        """Test error handling when LLM service fails"""
        service = LLMService()
        
        with pytest.raises(Exception) as exc_info:
            service.send_message("invalid_model", "Test message", {})
        
        assert "Unsupported model" in str(exc_info.value)

class TestWorkflowService:
    def test_create_workflow(self):
        """Test creating a new workflow"""
        service = WorkflowService()
        workflow_data = {
            "name": "Test Workflow",
            "description": "A test workflow",
            "steps": [
                {
                    "agent": "research_assistant",
                    "action": "research_topic",
                    "inputs": ["query"],
                    "outputs": ["research_data"]
                }
            ]
        }
        
        workflow = service.create_workflow(workflow_data)
        
        assert workflow.id is not None
        assert workflow.name == "Test Workflow"
        assert workflow.description == "A test workflow"
        assert len(workflow.steps) == 1
        assert workflow.steps[0]["agent"] == "research_assistant"
    
    def test_execute_workflow(self):
        """Test executing a workflow"""
        service = WorkflowService()
        
        # Create a simple workflow
        workflow_data = {
            "name": "Test Workflow",
            "steps": [
                {
                    "agent": "test_agent",
                    "action": "test_action",
                    "inputs": ["input_data"],
                    "outputs": ["output_data"]
                }
            ]
        }
        
        workflow = service.create_workflow(workflow_data)
        
        # Mock the agent service to avoid actual LLM calls
        with patch('backend.services.workflow_service.AgentService') as mock_agent_service:
            mock_agent = Mock()
            mock_agent.perform_action.return_value = {"output_data": "test_result"}
            mock_agent_service.return_value.get_agent.return_value = mock_agent
            
            result = service.execute_workflow(workflow.id, {"input_data": "test_input"})
            
            assert result["output_data"] == "test_result"
            mock_agent.perform_action.assert_called_once_with(
                "test_action", {"input_data": "test_input"}
            )
    
    def test_get_workflow_execution_history(self):
        """Test retrieving workflow execution history"""
        service = WorkflowService()
        
        # Create and execute a workflow
        workflow_data = {
            "name": "Test Workflow",
            "steps": [
                {
                    "agent": "test_agent",
                    "action": "test_action",
                    "inputs": ["input_data"],
                    "outputs": ["output_data"]
                }
            ]
        }
        
        workflow = service.create_workflow(workflow_data)
        
        with patch('backend.services.workflow_service.AgentService') as mock_agent_service:
            mock_agent = Mock()
            mock_agent.perform_action.return_value = {"output_data": "test_result"}
            mock_agent_service.return_value.get_agent.return_value = mock_agent
            
            service.execute_workflow(workflow.id, {"input_data": "test_input"})
            
            history = service.get_execution_history(workflow.id)
            
            assert len(history) == 1
            assert history[0]["workflow_id"] == workflow.id
            assert history[0]["status"] == "completed"